src/object.d

```diff
@@ -2790,7 +2790,54 @@ unittest
   +/
 void destroy(T)(T obj) if (is(T == class))
 {
-    rt_finalize(cast(void*)obj);
+    static if (__traits(hasMember, obj, "__dtor"))
```

Comment:
That cant work properly. destroy can be called for a class that has a nogc dtor but after downcasting.

Reply:
I thought at first you meant that `destroy()` might not be marked `@nogc` when a subclass is `@nogc`. This is by design, and the only possible way to do it. I then realized you might mean this:
```
class C {
    @nogc ~this() {}
    @nogc void foo() {}
}

class D : C {
    ~this() {}
    override void foo() {}
}

// tuple("@nogc", "@system")
pragma(msg, __traits(getFunctionAttributes, __traits(getMember, C, "__dtor")));
// tuple("@system")
pragma(msg, __traits(getFunctionAttributes, __traits(getMember, D, "__dtor")));
// tuple("@nogc", "@system")
pragma(msg, __traits(getFunctionAttributes, C.foo));
// tuple("@nogc", "@system")
pragma(msg, __traits(getFunctionAttributes, D.foo));
```
Clearly this is a hole in the type system - if a base function is marked `@nogc`, derived functions must be as well (and are, in the case of regular virtual functions).

Does this invalidate the PR? I disagree.

Reply:
class destructors are not virtual. That's a part of the problem. destructors are found and called using the TypeInfo_Class, not by inheritance, so

> derived functions must be as well

is unfortunately wrong. A real fix would involve huge changes. Try to work more on the problem and you'll see ;)

Reply:
I love being wrong - it means I get to learn something! :D

Looking at `rt_finalize2`, it seems the only issue is destructors - monitor deletion doesn't touch the GC. In other words, if `@nogc` on destructors were inherited, we could do the test I do in this PR, and it would give correct results (if a `@nogc` destructor is defined, it's safe to assume that all derived destructors are `@nogc`, in all other cases that's an unsafe assumption**). Of course, since it's not, it won't.

There may very well be something I've missed - are there any discussions on this topic that you can recommend? It seems weird to me that destructor attributes aren't considered part of the public interface of a class, but I can accept there being reasons for it - I just don't know what they are.

** Actually, for final classes with no destructor, it would also be safe.

Reply:
>  are there any discussions on this topic that you can recommend?

No.  i see this problem as a dead end. In first place constructor and destructor not being virtual is a fundamental design error imo. likely unfixable.

Reply:
There is actually a subset of classes for which destroy can safely be `@nogc` today: final classes where all fields and base classes define `@nogc` destructors or no destructors at all.